content-src/lib/selectLayoutRender.js

```diff
@@ -38,6 +38,34 @@ export const selectLayoutRender = (state, rickRollCache) => {
   }
 
   const positions = {};
+  const filterComponentsByPref = () => {
+    const DS_COMPONENTS = ["Message", "SectionTitle", "Navigation",
+      "CardGrid", "Hero", "HorizontalRule", "List"];
+
+    const filterArray = [];
+
+    if (!prefs["feeds.topsites"]) {
+      filterArray.push("TopSites");
+    }
+
+    if (!prefs["feeds.section.topstories"]) {
+      filterArray.push(...DS_COMPONENTS);
+    }
+
+    for (const [rowIndex, row] of Object.entries(layout)) {
+      for (const [index, component] of Object.entries(row.components)) {
+        if (filterArray.includes(component.type)) {
+          if (row.components.length === 1) {
+            layout.splice(rowIndex, 1);
```

Comment:
I used a double for loop for #4890 because I wanted to return early. Might be cleaner/safer to use array methods, e.g., reduce/filter instead of modifying the array that you're iterating over as we need to visit every item anyway.

Reply:
E.g.,
```js
return layout.map(row => ({
    ...row,
      components: row.components.filter(component => !filterArray.includes(component.type))
  })).filter(row => row.components.length).map(row => ({
    ...row,
```
And no need to have a separate `filterComponentsByPref` inline function as it only gets called once.

Reply:
Actually, looking at this again, the existing `layout.map` looks pretty much like what we want to do anyway. So perhaps more like

```js
return layout.map(row => ({
  ...row,
  // Loops through desired components and adds a .data property
  components: row.components.filter(c => !filterArray.includes(c.type)).map(component => …),
})).filter(row => row.components.length);

So just before adding .data property, first filter out the components. Then if there's no more components in a row, filter out empty rows.

Reply:
This should also make coverage happy as there aren't separate conditional branches to test.